[notifier] extend otChangeFlags to uint64_t and add epskc mode change event

bukepo opened the pull request.

abtink commented:
Thanks @bukepo. Updating `otChangeFlags` to `uint64_t` sounds reasnable when/if we need new events.

Regarding the addition of the "epskc change event", the BA already provides a callback `otBorderAgentEphemeralKeyCallback` for this. Is there any reason to add this to the `Notifier` as well?  I would personally prefer to provide one way for this.

abtink commented on src/core/common/notifier.cpp:
```diff
@@ -233,7 +235,7 @@ void Notifier::LogEvents(Events aEvents) const
 
 exit:
-    LogInfo("StateChanged (0x%08lx) %s%s]", ToUlong(aEvents.GetAsFlags()), didLog ? "... " : "[", string.AsCString());
+    LogInfo("StateChanged (0x%" PRIx64 ") %s%s]", aEvents.GetAsFlags(), didLog ? "... " : "[", string.AsCString());
```
I strongly suggest not using `PRIx` definitions. While they are intended to help with portability across platforms and toolchains, ironically they often cause more issues than they solve. 

Some platforms/toolchains may not implement or provide them correctly. I've encountered issues in the past (in other projects) where incorrect definitions were provided by embedded toolchains, leading to subtle and hard-to-debug bugs. We faced some earlier issues in OT as well:  https://github.com/openthread/openthread/issues/5554

On a style note, I personally find them making code less readable and ugly.

bukepo commented:
> Thanks @bukepo. Updating `otChangeFlags` to `uint64_t` sounds reasnable when/if we need new events.
> 
> Regarding the addition of the "epskc change event", the BA already provides a callback `otBorderAgentEphemeralKeyCallback` for this. Is there any reason to add this to the `Notifier` as well? I would personally prefer to provide one way for this.

We have the pskc changed event, adding the epskc changed event can keep OpenThread APIs consistent. Perhaps we should deprecate `otBorderAgentEphemeralKeyCallback` which was added when there's no available bit for the epskc changed event. Thoughts?

abtink commented:
> We have the pskc changed event, adding the epskc changed event can keep OpenThread APIs consistent. Perhaps we should deprecate otBorderAgentEphemeralKeyCallback which was added when there's no available bit for the epskc changed event. Thoughts?

Regarding events, I wouldn't consider PSKc and ephemeral key to be similar.

PSKc is a main parameter of the Thread network (and included in the Operational Dataset). We provide APIs to get the current PSKc, so having an event to signal when it changes seems reasonable. 

On the other hand, an ephemeral key is by definition temporary, and one-time use. We provide APIs to set it but not one to get the current value (the user shouldn't need to know that). The choice of using a callback for the ephemeral key and not using the `Notifier` was intentional. The callback signals more than just when the ephemeral key is set; it signals when any of the related parameters (port number, timeout) is changed, and/or when its use is finished (commissioner disconnected, multiple failed connection attempts, timeouts, etc.).

bukepo commented:
> > We have the pskc changed event, adding the epskc changed event can keep OpenThread APIs consistent. Perhaps we should deprecate otBorderAgentEphemeralKeyCallback which was added when there's no available bit for the epskc changed event. Thoughts?
> 
> Regarding events, I wouldn't consider PSKc and ephemeral key to be similar.
> 
> PSKc is a main parameter of the Thread network (and included in the Operational Dataset). We provide APIs to get the current PSKc, so having an event to signal when it changes seems reasonable.
> 
> On the other hand, an ephemeral key is by definition temporary, and one-time use. We provide APIs to set it but not one to get the current value (the user shouldn't need to know that). The choice of using a callback for the ephemeral key and not using the `Notifier` was intentional. The callback signals more than just when the ephemeral key is set; it signals when any of the related parameters (port number, timeout) is changed, and/or when its use is finished (commissioner disconnected, multiple failed connection attempts, timeouts, etc.).

Thanks for explaining the intention. You are right, `otBorderAgentEphemeralKeyCallback` callback not only notifies for ephemeral key changes, but also other parameter changes. From the code, there are two places where the callback is posted, and it looks like we can use the notifier to do the same job, but saving the additional callback function pointer, callback context and the `EphemeralKeyTask`.

I updated the code to notify it just as the existing callback. I also renamed the event name to `epskc mode`, indicating it's beyond the ephemeral key itself. Could you take another look?

librasungirl commented:
> > We have the pskc changed event, adding the epskc changed event can keep OpenThread APIs consistent. Perhaps we should deprecate otBorderAgentEphemeralKeyCallback which was added when there's no available bit for the epskc changed event. Thoughts?
> 
> Regarding events, I wouldn't consider PSKc and ephemeral key to be similar.
> 
> PSKc is a main parameter of the Thread network (and included in the Operational Dataset). We provide APIs to get the current PSKc, so having an event to signal when it changes seems reasonable.
> 
> On the other hand, an ephemeral key is by definition temporary, and one-time use. We provide APIs to set it but not one to get the current value (the user shouldn't need to know that). The choice of using a callback for the ephemeral key and not using the `Notifier` was intentional. The callback signals more than just when the ephemeral key is set; it signals when any of the related parameters (port number, timeout) is changed, and/or when its use is finished (commissioner disconnected, multiple failed connection attempts, timeouts, etc.).

I saw @bukepo has renamed the event as `kEventEpskcModeChanged` for accuracy, one benefit of adding such a new event is to allow more than one watchers, by leveraging existing mechanism, e.g. allowing `ot-ctl ba ephemeralkey callback enable`  and otbr's epskc functionality at the same time.

abtink commented:
Thanks guys!

I'm uncertain about this change. We already have the `otBorderAgentEphemeralKeyCallback` mechanism in place.
- Generally, our policy regarding existing OT APIs is to keep them unchanged to ensure backward compatibility.
- In this case, the e-PSKC mechanism is a key new feature. There might be companies/vendors developing this feature and integrating OT source code with existing APIs. Removing these APIs (and replacing them with a notifier event) could create issues for developers using them.

Considering this, we cannot and should not remove the existing callback mechanism and its API.

----

The question then becomes, **do we want to add `Notifier` events in addition to the callback mechanism?**

My preference is to provide one way to achieve this. Having multiple similar ways just adds to user confusion.

I understand Rongli's arguments about allowing multiple watchers for these changes. My view is that having a single callback is actually better. While it may seem easier to allow multiple watchers, I believe a single callback leads to better overall software design:

- Ideally, there should be one software entity (as part of the application running over the OT stack) that manages the e-PSKc behavior and calls OT APIs to set and clear the key. We don't want multiple entities to be able to set the ephemeral key.
- This same entity can use the single OT callback to be notified about changes and take appropriate actions, it can then signal other modules (if needed).
- This model follows the layered design principle and encourages a better software architecture.
- Allowing multiple watchers of this event (which may then take different actions) can lead to subtle bugs and issues. 

Regarding `ot-ctl ba ephemeralkey callback`, it is intended for unit testing only. It functions similarly to other `cli` commands that can be used to override parameters/actions by the `otbr` modules (e.g., `channel 12` or `ba ephemeralkey <key>` can mess up previous configs set by `ot-br-posix`). We can consider removing this CLI command entirely to avoid confusion.

bukepo commented:
I acknowledge that epskc is a **key new** feature. I personally think the impact of switching to the notifier style would be relatively small for the following reasons:

* We are still maturing the feature in OpenThread, I don't expect many vendors have completed integration of it as even OpenThread hasn't finished the integration in ot-br-posix yet.
* For those already integrated with the current API, adapting to the notifier style should be a trivial effort since the notifier is not a completely new thing, but existing OpenThread mechanism for other state changes.

---

Besides, it may also be confusing to developers why some state changes use the notifier, some not. Having consistent APIs eases the integration of this feature in the long term. The cost of doing it now should be much smaller than in the future, so I think we still have a good opportunity to make it. Thoughts?

abtink commented:
> Besides, it may also be confusing to developers why some state changes use the notifier, some not. Having consistent APIs eases the integration of this feature in the long term. 

@bukepo , there is a clear distinction here.

`Notifier` (`otStateChangedCallback`) is generally used to signal changes to operational configuration of the Thread Network. These represent a consistent state or parameter of Thread network. Examples include device role, operational dataset, network key, network data, IPv6 addresses, etc. 

Callbacks are used (and preferred) to signal events tied to specific (possibly transient) actions triggered by API calls.  The user initiates an action (invoking an API), and its related events are signaled via a registered callback. There are many existing examples of this pattern in OT:

  * `otDetachGracefullyCallback`: Triggered after graceful detach API.
  * `otHandleActiveScanResult`: After scan-related APIs are called.
  * `otSrpClientCallback`: To signal events related to SRP registrations
  * `otSrpServerServiceUpdateHandler`: To signal changes to the SRP server.
  * `otHandleCoapSecureClientConnect`: To signal CoAP Secure connected after a connection request.

The ephemeral ket fits the second pattern, making the callback model preferable. It's a transient action triggered by an explicit API call, not a general state of the Thread network. Using `Notifier` events for it doesn't make sense to me.

bukepo commented:
> > Besides, it may also be confusing to developers why some state changes use the notifier, some not. Having consistent APIs eases the integration of this feature in the long term.
> 
> @bukepo , I believe there is a clear distinction here.
> 
> `Notifier` (`otStateChangedCallback`) is generally used to signal changes to operational configuration of the Thread Network. These represent a consistent state or parameter of Thread network. Examples include device role, operational dataset, network key, network data, IPv6 addresses, etc.
> 

I don't quite get the meaning of *consistent state or parameter*, and why espkc mode doesn't belong to them. We enable epskc mode, and it's consistently active for a while. I don't see fundamental differences from other states like the device role, border agent state, on-mesh commissioner state, etc. The device role is runtime determined based on not only the device's own capability/configuration, but also the network topology. The device role can change to the leader, router, end device or detached automatically. The epskc mode is also similar to the on-mesh commissioner state, which can be enabled and disabled temporarily.

> Callbacks are used (and preferred) to signal events tied to specific (possibly transient) actions triggered by API calls. The user initiates an action (invoking an API), and its related events are signaled via a registered callback. There are many existing examples of this pattern in OT:
> 
> * `otDetachGracefullyCallback`: Triggered after graceful detach API.
> * `otHandleActiveScanResult`: After scan-related APIs are called.
> * `otSrpClientCallback`: To signal events related to SRP registrations
> * `otSrpServerServiceUpdateHandler`: To signal changes to the SRP server.
> * `otHandleCoapSecureClientConnect`: To signal CoAP Secure connected after a connection request.
> 
> The ephemeral key fits the second pattern, making the callback model preferable. It's a transient action triggered by an explicit API call, not a general state of the Thread network. Using `Notifier` events for it doesn't make sense to me.

I fully agree with you that these callbacks you listed are different from the state changes. But I still believe the epskc mode is more close to state changes.

abtink commented:
> I don't quite get the meaning of consistent state or parameter, and why espkc mode doesn't belong to them. 

I mean these are parameters/states of Thread network that are "always" present and well-defined and we provide APIs to retrieve them at any time. We can always get the current device role (even though the role itself may change). The same applies to other parameters; we can always get the Active Dataset, Network Data, list of IPv6 addresses, etc. Even in the case of PSKc, we can always ask and get the current PSKc (even when an ephemeral key is in use). 

In contrast, an ephemeral key is, by its very nature, short-lived and temporary. It requires an explicit API call to be used and isn't stored within `BorderAgent`. We also don't have an API to fetch the current ephemeral key.

The callback serves to notify about changes not only in the ephemeral key but also in its related parameters (such as port and timeout) or disconnect status. All of this is documented in the callback definition. I understand you renamed the event to `OT_CHANGED_EPSKC_MODE`, but the term "ePSKc mode" itself isn't a clearly defined state or parameter. It's a collection of related events and parameters.

bukepo commented:
> I mean these are parameters/states of Thread network that are "always" present and well-defined and we provide APIs to retrieve them at any time. We can always get the current device role (even though the role itself may change). The same applies to other parameters; we can always get the Active Dataset, Network Data, list of IPv6 addresses, etc. Even in the case of PSKc, we can always ask and get the current PSKc (even when an ephemeral key is in use).
> 
> In contrast, an ephemeral key is, by its very nature, short-lived and temporary. It requires an explicit API call to be used and isn't stored within `BorderAgent`. We also don't have an API to fetch the current ephemeral key.
> 
> The callback serves to notify about changes not only in the ephemeral key but also in its related parameters (such as port and timeout) or disconnect status. All of this is documented in the callback definition. I understand you renamed the event to `OT_CHANGED_EPSKC_MODE`, but the term "ePSKc mode" itself isn't a clearly defined state or parameter. It's a collection of related events and parameters.

Thank you for the explanation. I understand the differences you see now. May I know the consideration differentiating them? It looks some existing events are also not so consistent such as address added/removed event. We don't have an API to fetch the newly added/removed address either.

I think the main purpose of the notifier is to notify state subscribers that some states of OpenThread changed so that subscribers can take some actions. Whether the states are consistent or not doesn't harm the goal. Whether it's a collection or a singular doesn't harm the goal(Active/Pending dataset are actually a collection of parameters).

We may define `ePSKc mode` changed event as what the current `otBorderAgentEphemeralKeyCallback` notifies, like `Any parameter related to the ephemeral key changed`. What do you think?

abtink commented:
> We may define ePSKc mode changed event as what the current otBorderAgentEphemeralKeyCallback notifies, like Any parameter related to the ephemeral key changed. What do you think?

Let me offer another example to illustrate my point. To me, an ephemeral key functions similarly to a **Mac/Discover scan**. We call an API to initiate the scan, provide a callback, receive the results, and ultimately get a callback indicating the scan's completion. Similarly, for an ephemeral key, we request it and receive a callback when it starts or its parameters change, and get a final callback when it's finished. 

It doesn't make logical sense to define a "scan mode" event and use a `Notifier` to signal the end of a scan. Likewise, defining "ePSKc mode" as a network parameter feels forced when it's intrinsically a combination of parameters and events, and it represents a short-lived action.

Furthermore, I see no compelling reason to replace the existing, functional callback model with a new `Notifier`-based one.

Conversely, I've previously argued how using a callback model leads to better software architecture and enforces a layered model (avoiding cross-layer eventing). https://github.com/openthread/openthread/pull/10642#issuecomment-2316100807

bukepo commented:
> Let me offer another example to illustrate my point. To me, an ephemeral key functions similarly to a **Mac/Discover scan**. We call an API to initiate the scan, provide a callback, receive the results, and ultimately get a callback indicating the scan's completion. Similarly, for an ephemeral key, we request it and receive a callback when it starts or its parameters change, and get a final callback when it's finished.
> 
> It doesn't make logical sense to define a "scan mode" event and use a `Notifier` to signal the end of a scan. Likewise, defining "ePSKc mode" as a network parameter feels forced when it's intrinsically a combination of parameters and events, and it represents a short-lived action.
> 
> Furthermore, I see no compelling reason to replace the existing, functional callback model with a new `Notifier`-based one.
> 
> Conversely, I've previously argued how using a callback model leads to better software architecture and enforces a layered model (avoiding cross-layer eventing). [#10642 (comment)](https://github.com/openthread/openthread/pull/10642#issuecomment-2316100807)

I fully agree with you that it doesn't make sense using notifier to collect *scan results*. The epskc state is not something similar to Mac/Discover scan results to me. Because scan results are **transient** data and need to be collected synchronously and notifier is asynchronous. If we do care some "scan mode" some day, it make sense to me to define a "scan mode" and notify whether the device is in "scan mode"(perhaps to tell the upper layer to restrain from sending packets at the moment). 

There's no compelling reason to replace the callback model. The motivation is to move to the existing general Notifier model for consistency, and also to pave the way for other states like PD.

Could you be more specific how this callback model leads to better software architecture? It looks both notifier and the callback is on the interface of OpenThread. The notifier is not a cross layer event system(such as Qt's Signals/Slots, NodeJS's event emitter) that could spread OpenThread events to other layers.

abtink commented:
@bukepo did you notice the increase in code size +592/+762. Where is the increase coming from? This is obviously concerning.

|  name  |  branch  |  text  | data  | bss  | total |
| :----: | :------: | -----: | ----: | ---: | ----: |
| ot-cli-ftd | main | 469320 | 856 | 66484 | 536660 |
|  | #10642 | 469912 | 856 | 66492 | 537260 |
|  | +/- | +592 | 0 | +8 | +600 |
| ot-ncp-ftd | main | 437900 | 760 | 61712 | 500372 |
|  | #10642 | 438660 | 760 | 61720 | 501140 |
|  | +/- | +760 | 0 | +8 | +768 |

abtink commented:
> Could you be more specific how this callback model leads to better software architecture?

Already explained this. Please check the comment https://github.com/openthread/openthread/pull/10642#issuecomment-2316100807.
